### DISCUSSION TITLE
Excluding parameter as structural

### DISCUSSION THREAD

**Discussion by Arinomo** [OpenModelica/OpenModelica#8936] [Q&A]
consider following MWE

```.mo
model structuralTest
constant Real a = 5.0;
parameter Real x = 2.0;
parameter Real y = 3.0;

Real z;
equation

if x < a then
  z = 15 *y;
else
  z = 10 * y;
end if;

end structuralTest;
```

the parameter `x` is a structural parameter and could no be modified without re-building the model. It is not a problem when working with OMEdit or OMPython, but when one exports above model as FMU, It will be hard to modified the value of `x`. Is it possible to exclude a parameter as a structural parameter when compiling a model?

**Comment by adrpo** [comment]
The `x` parameter should not be structural in this example. Only when you have unbalanced if equations (more or none equations in one branch) then it should be marked as structural. 

@perost hopefully our compiler does not mark `x` as structural here.

**Reply by perost** [reply]
The new frontend currently does mark parameter conditions in if-equations as structural, because that's what the old frontend did and we get some issues with the backend if we don't. But I will investigate if it's still needed, and if it's not possible to change right now then maybe we can add a compiler flag to allow users to disable it.

**Reply by Arinomo** [reply]
> Only when you have unbalanced if equations (more or none equations in one branch) then it should be marked as structural.

Does this behavior also true for when-clause like:

```.mo
when  x < a then
  z := 15 *y;
end when;
```
(i.e. ```x``` will always be structural since there is no other branch)?

**Reply by perost** [reply]
No, when-conditions are never evaluated by the frontend.

**Reply by adrpo** [reply]
When equations are instantaneous, they are active only at events so it does not matter in this case. I fail to see what a parameter condition in an when equation would be good for.

**Reply by Arinomo** [reply]
additional info: 
if i done this
```.mo
model structuralTest
constant Real a = 5.0;
parameter Real x = 2.0;
parameter Real y = 3.0;

Real z;
algorithm

if x < a then
  z := 15 *y;
else
  z := 10 * y;
end if;

end structuralTest;
```
then the paramter ```x``` is no longer considered structural.

> I fail to see what a parameter condition in an when equation would be good for.

I don't know why i asked about when-clause, i was thinking about for-loop actually (i guess today is my zoned out day).

```.mo
parameter Interger a = 3;
parameter Real x = 2.0;
Real z[N];

equation
for i in 1:N loop
    z[i] = x * i;
end for;
```

will always evaluate ```N``` as structural parameter right?

**Reply by perost** [reply]
Yeah, we don't evaluate conditions in if-algorithms, so they are not considered structural.

For for-loops we normally unroll them in equation sections, unless you use the experimental flag `-d=-nfScalarize` (i.e. turn off scalarization). But turning off scalarization has other consequences too, and I'm not sure how well it would work with FMI.

In that particular example we would consider `N` to be structural anyway though, since it's used to define the dimension of `z`.

**Comment by casella** [comment]
I agree. In many cases you actually don't want to change such parameters at run time, and you'd expect the compiler to only generate code for the selected branch of the if-equation. That has been the behaviour so far, so I'm in favour of relaxing it under some flag.

In fact this is a grey area, because the Modelica Specification actually doesn't mandate a specific behaviour in these cases, so it's mostly a tool issue how to handle them.

**Reply by Arinomo** [reply]
Yes, i think relaxing it via flag would be great. Since more often than not, we test a FMU on other system and the work to change a parameter then export it again adds up

**Comment by perost** [comment]
I played around with not evaluating if-conditions, but it's a bit complicated and hard to change without breaking things. So instead I implemented #8990 which lets you use the Evaluate annotation to tell the compiler that a parameter should not be evaluated (previously we only handled the opposite, i.e. Evaluate=true). Hopefully this solves your issue.

So if you now change the definition of `x` to:
```Modelica
  parameter Real x = 2.0 annotation(Evaluate = false);
```
then the if-condition will not be evaluated.

Also note that this is just a suggestion to the compiler and will be ignored in some cases (such as when a parameter is used to define the size of an array variable).

**Reply by Arinomo** [reply]
Thanks @perost, i will try it when its available on the nightly

**Reply by casella** [reply]
Thanks @perost, in fact this seems to me a brilliant solution, because it leaves to the modeller the choice of how to handle the code generation.
